data/config.yaml

```diff
@@ -4,6 +4,7 @@
 
 production: &defaults
     maxBlockSize: 1000000 # 1 MB
+    mutableChainSize: 2016 # blocks
```

Comment:
What do you have in mind exactly? The way I see things, you do want to have two kinds of configurations: one which is protocol specific and that can travel on-chain, and some part which is off-chain. While we may argue that `maxBlockSize` could have lived on chain, I suppose `mutableChainSize` is something we might want to leave in a separate config file.

Perhaps as part of a more luxurious, more decorated shed, we might argue that we might want to bake everything into the initial genesis file. This might feel abstruse but it's how some projects do that (my experience is only limited to Cardano obviously), i.e. [genesis.json](https://raw.githubusercontent.com/input-output-hk/cardano-sl/develop/lib/mainnet-genesis.json) (⚠️, very big file). It basically contained the money for the initial stakeholders, the start time of the chain, the certificates for the nodes, the _delegation_ certs (PoS stuff) and the "protocol configuration params", which is something at the moment we don't bake into genesis (but it lives in this `config.yaml` separate file).

Reply:
Where I'm coming from is this: in the not-so-distant future, we want to deploy this application to a bunch of servers, preferably continuously. For the latter to work out, we want to make sure that the build catches as many issues as possible which could cause a faulty deployment. An often overlooked aspect of this is that the _configuration_ needs to be correct, too. If the configuration is scattered around a bunch of semi-structured text files, magic environment variables, _and_ a set of CLI options (some of which having default values), this actually becomes a rather hairy problem.

One solution, as you suggest, is to stick *everything* into a single config file, which is parsed into a strongly typed representation + some runtime validation if necessary. An actual configuration - either checked in to source control, or generated on the fly as part of the build - can now be tested by running the parser and validation _without running the application itself_. 

The other option I've used in the past is to stick *everything* into CLI options. `optparse-applicative`, for all it's quirks, is actually quite composable, and a given `Parser` can be evalutated purely. The advantage of this is that the configuration is always explicit (modulo default values), and visible even at runtime (just `ps`). A drawback is that secret configuration values have to still be resolved from files or via network calls, but that problem doesn't really go away with the mono-config-file approach (the file has to be rewritten at runtime).

As for this particular config file, it is not so clear if and when we would actually want to supply a different configuration from the outside. Why can't this be a function `defaultConfig :: Environment -> Config` in the code?

Reply:
Right, thanks for explaining the rationale!

> As for this particular config file, it is not so clear if and when we would actually want to supply a different configuration from the outside. Why can't this be a function defaultConfig :: Environment -> Config in the code?

I suppose we totally can -- if we don't care about the virtue of having an amendable config file which can be modified without recompiling and releasing a new executable, for sure we can. Happy to go with what the team decides!

Reply:
CLI-only is a good option in my opinion.

One thing I'd like to bring up though is that stuff like `maxBlockSize` and `mutableChainSize` *shouldn't* be configurable so easily. I think this should be a static variable in the source code. The reason being simply that it can break consensus if changed. In my opinion of course, "configuration" should be things you can change on your node without too much risk.

Reply:
> shouldn't be configurable so easily[...] I think this should be a static variable in the source code.

Or, even better, stored on chain 😁 Without entering a bikeshed round, should I simply bit the bullet and add an extra commit to this (already gigantic) PR and simply scrap the read-from-file `Config` in favour of a fully static `Config`? I don't think it's going to be invasive as the external interface wouldn't change, only the fact now in `exe/Main.hs` we don't need to do IO to acquire a config. 👍 ? 👎  ?

Reply:
Yeah, I think it's a good idea. If we want it configurable later we can add a CLI flag, but for now it can be static.